PR: https://github.com/odoo/odoo/pull/223685

From: 72fffcac7720f857e90ded7c6a204c114d4630b2
From: Arnold Moyaux
Date: 2025-09-09 18:25:14

Breaking data model changes scores: del:2 + add:0, change matches:
-    route_ids = fields.Many2many(default=lambda self: self._get_buy_route())

Total Changes: 58

[IMP] stock: buy and mnaufacture as warehouse route

The purpose is to simplify the configuration of products.
You just need to create a BoM or add a vendor to make it
ready to produce/sell.

However it works as before, so people that want to play
with priority or have a complex use case can reswitch the
rule to product.

task-4385221

closes odoo/odoo#223685

Related: odoo/enterprise#92820
Related: odoo/upgrade#8256
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/purchase_stock/models/product.py
+++ b/addons/purchase_stock/models/product.py
@@ -7,19 +7,25 @@ from odoo import api, fields, models
 from odoo.fields import Domain
 from odoo.tools import formatLang
 from odoo.tools.float_utils import float_round
+from odoo.exceptions import UserError
 
 
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
-    @api.model
-    def _get_buy_route(self):
-        buy_route = self.env.ref('purchase_stock.route_warehouse0_buy', raise_if_not_found=False)
-        if buy_route:
-            return self.env['stock.route'].search([('id', '=', buy_route.id)]).ids
-        return []
+    @api.constrains('route_ids', 'purchase_ok')
+    def _check_buy_route(self):
+        non_purchasable_products = self.filtered(lambda product: not product.purchase_ok)
+        if not non_purchasable_products:
+            return
 
-    route_ids = fields.Many2many(default=lambda self: self._get_buy_route())
+        buy_routes = self.env['stock.rule'].search([
+            ('action', '=', 'buy'),
+            ('picking_type_id.code', '=', 'incoming'),
+            ('active', '=', True),
+        ]).route_id
+        if buy_routes and any(buy_routes & product.route_ids for product in non_purchasable_products):
+            raise UserError(self.env._("The 'Buy' route cannot be assigned to a product that is not purchasable. Enable 'Can be Purchased' boolean to use this route."))
 
 
 class ProductProduct(models.Model):

--- a/addons/purchase_stock/models/stock.py
+++ b/addons/purchase_stock/models/stock.py
@@ -3,7 +3,7 @@
 from collections import defaultdict
 from dateutil import relativedelta
 
-from odoo import api, fields, models, _
+from odoo import api, Command, fields, models, _
 from odoo.fields import Domain
 
 
@@ -41,10 +41,35 @@ class StockPicking(models.Model):
 class StockWarehouse(models.Model):
     _inherit = 'stock.warehouse'
 
-    buy_to_resupply = fields.Boolean('Buy to Resupply', default=True,
-                                     help="When products are bought, they can be delivered to this warehouse")
+    buy_to_resupply = fields.Boolean(
+        'Buy to Resupply', compute='_compute_buy_to_resupply',
+        inverse="_inverse_buy_to_resupply", default=True,
+        help="When products are bought, they can be delivered to this warehouse")
     buy_pull_id = fields.Many2one('stock.rule', 'Buy rule', copy=False)
 
+    def _compute_buy_to_resupply(self):
+        for warehouse in self:
+            buy_route = warehouse.buy_pull_id.route_id
+            warehouse.buy_to_resupply = bool(buy_route.product_selectable or buy_route.warehouse_ids.filtered(lambda w: w.id == warehouse.id))
+
+    def _inverse_buy_to_resupply(self):
+        for warehouse in self:
+            buy_route = warehouse.buy_pull_id.route_id
+            if not buy_route:
+                buy_route = self.env['stock.rule'].search([
+                    ('action', '=', 'buy'), ('warehouse_id', '=', warehouse.id)]).route_id
+            if warehouse.buy_to_resupply:
+                buy_route.warehouse_ids = [Command.link(warehouse.id)]
+            else:
+                buy_route.warehouse_ids = [Command.unlink(warehouse.id)]
+
+    def _create_or_update_route(self):
+        purchase_route = self._find_or_create_global_route('purchase_stock.route_warehouse0_buy', _('Buy'))
+        for warehouse in self:
+            if warehouse.buy_to_resupply:
+                purchase_route.warehouse_ids = [Command.link(warehouse.id)]
+        return super()._create_or_update_route()
+
     def _generate_global_route_rules_values(self):
         rules = super()._generate_global_route_rules_values()
         location_id = self.lot_stock_id
@@ -220,6 +245,13 @@ class StockWarehouseOrderpoint(models.Model):
 
         return result
 
+    def _filter_warehouse_routes(self, product, warehouses, route):
+        if any(rule.action == 'buy' for rule in route.rule_ids):
+            if product.seller_ids:
+                return super()._filter_warehouse_routes(product, warehouses, route)
+            return False
+        return super()._filter_warehouse_routes(product, warehouses, route)
+
     def _get_default_route(self):
         route_ids = self.env['stock.rule'].search([
             ('action', '=', 'buy')
